src/nvim/event/rstream.c

```diff
@@ -167,9 +169,19 @@ static void fread_idle_cb(uv_idle_t *handle)
   invoke_read_cb(stream, false);
 }
 
-static void invoke_read_cb(Stream *stream, bool eof)
+static void read_event(void **argv)
 {
+  Stream *stream = argv[0];
+  bool eof = (uintptr_t)argv[1];
   if (stream->read_cb) {
     stream->read_cb(stream, stream->buffer, stream->data, eof);
   }
+  if (eof && stream->internal_eof_cb) {
+    stream->internal_eof_cb(stream, stream->internal_data);
+  }
+}
+
+static void invoke_read_cb(Stream *stream, bool eof)
+{
+  queue_put(stream->events, read_event, 2, stream, (void *)(uintptr_t)eof);
```

Comment:
>Is there any other scenario where you can see this reordering would cause a problem?

test2.sh:
```bash
#!/bin/bash

echo "stdout 1"
echo "stderr 1"  1>&2
echo "stdout 2"
echo "stderr 2"  1>&2
```
Output for test2.sh in a terminal:
```bash
stdout 1
stderr 1
stdout 2
stderr 2
```
Now with a vim script:
```vim
function! s:JobHandler(job_id, data, event)
    call append(line('$'),a:data)
endfunction

let s:callbacks = {
\ 'on_stdout': function('s:JobHandler'),
\ 'on_stderr': function('s:JobHandler'),
\ }
let job1 = jobstart("sh ./test2.sh", s:callbacks)
```
Output:
```vim

stdout 1
stdout 2

stderr 1
stderr 2


```
Now output is printed out of order.